src/stan/variational/advi.hpp

```diff
@@ -94,15 +89,30 @@ namespace stan {
        * @return   evidence lower bound (elbo)
        */
       double calc_ELBO(const Q& variational) const {
+        static const char* function =
+          "stan::variational::advi::calc_ELBO";
+
         double elbo = 0.0;
+        double energy_i;
         int dim = variational.dimension();
-
         Eigen::VectorXd zeta(dim);
+        int n_monte_carlo_drop = 0;
 
         for (int i = 0; i < n_monte_carlo_elbo_; ++i) {
           zeta = variational.sample(rng_);
-          // Accumulate log probability
-          elbo += model_.template log_prob<false, true>(zeta, print_stream_);
+          try {
+            energy_i = model_.template log_prob<false, true>(zeta,
+                                                             print_stream_);
+            stan::math::check_not_nan(function, "energy_i", energy_i);
+            stan::math::check_finite(function, "energy_i", energy_i);
+            elbo += energy_i;
+          } catch (std::exception& e) {
+            this->write_error_msg_(print_stream_, e);
+            i -= 1;
```

Comment:
Ack --- Dustin's already changed this, which is why I'm not seeing
what Daniel was objecting to any more.  I think the rewrite's going
in the right direction.  I had a few more concerns:

* check_finite should catch NaN, so I don't think you also need check_not_nan

* variables should be declared as close to where they're used as possible,
so the energy_i variable should be down in the loop with

  double energy_i
    = model_.template ...

* the zeta is an exception because you don't want to have to create
new vectors.  That's why you probably want to pass the vector
zeta as a non-constant reference argument to variational.sample(rng_, zeta)
so that the sample() method can fill it --- otherwise you'll be creating
and discarding vectors anyway and having to do a loop of the dimensionality
for assignment


        double elbo = 0.0;
        double energy_i;
        int dim = variational.dimension();
        Eigen::VectorXd zeta(dim);

        int i = 0;
        int n_monte_carlo_drop = 0;
        while (i < n_monte_carlo_elbo_) {
          zeta = variational.sample(rng_);
          try {
            energy_i = model_.template log_prob<false, true>(zeta,
                                                             print_stream_);
            stan::math::check_not_nan(function, "energy_i", energy_i);
            stan::math::check_finite(function, "energy_i", energy_i);
            elbo += energy_i;
            i += 1;
          } catch (std::exception& e) {
            this->write_error_msg_(print_stream_, e);
            n_monte_carlo_drop += 1;
            stan::math::check_less(function, "n_monte_carlo_drop",
              n_monte_carlo_drop, n_monte_carlo_elbo_);
          }
        }

- Bob



> On Jul 29, 2015, at 5:12 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> >  
> >          for (int i = 0; i < n_monte_carlo_elbo_; ++i) {
> >            zeta = variational.sample(rng_);
> > -          // Accumulate log probability
> > -          elbo += model_.template log_prob<false, true>(zeta, print_stream_);
> > +          try {
> > +            energy_i = model_.template log_prob<false, true>(zeta,
> > +                                                             print_stream_);
> > +            stan::math::check_not_nan(function, "energy_i", energy_i);
> > +            stan::math::check_finite(function, "energy_i", energy_i);
> > +            elbo += energy_i;
> > +          } catch (std::exception& e) {
> > +            this->write_error_msg_(print_stream_, e);
> > +            i -= 1;
> 
> instead of this, which takes a minute to process what you're doing, could you do something like:
> 
> n = 0;
> int n_monte_carlo_drop = 0;
> while (n < n_monte_carlo_elbo_ && n_monte_carlo_drop < n_monte_carlo_elbo_) {
> ...
> }
> 
> Or something like that. Decrementing the loop index within the loop is just tricky to figure out.
> 
> —
> Reply to this email directly or view it on GitHub.
>